src/main/java/net/minecraftforge/client/ClientCommandHandler.java

```diff
@@ -0,0 +1,312 @@
+    {
+        for (CommandNode<S> child : sourceNode.getChildren())
+        {
+            ArgumentBuilder<S, ?> builder = child.createBuilder();
+            CommandNode<S> copy = builder.build();
+            copy(child, copy);
+            resultNode.addChild(copy);
+        }
+    }
+
+    /**
+     * 
+     * Deep copies the children of a command node and stores a link between the source and the copy
+     * 
+     * @param sourceNode
+     *            the original command node
+     * @param resultNode
+     *            the result command node
+     * @param sourceToResult
+     *            a map storing the original command node as the key and the result command node as the value
+     * @param canUse
+     *            used to check if the player can use the command
+     * @param execute
+     *            the command to execute in place of the old command
+     * @param sourceToResultSuggestion
+     *            a function to convert from the {@link SuggestionProvider} with the original source stack to the {@link SuggestionProvider} with the result source stack
+     */
+    private static <S, T> void mergeCommandNode(CommandNode<S> sourceNode, CommandNode<T> resultNode, Map<CommandNode<S>, CommandNode<T>> sourceToResult,
+            S canUse, Command<T> execute, Function<SuggestionProvider<S>, SuggestionProvider<T>> sourceToResultSuggestion)
+    {
+        sourceToResult.put(sourceNode, resultNode);
+        for (CommandNode<S> sourceChild : sourceNode.getChildren())
+        {
+            if (sourceChild.canUse(canUse)) 
+            {
+                resultNode.addChild(toResult(sourceChild, sourceToResult, canUse, execute, sourceToResultSuggestion));
+            }
+        }
+    }
+
+    /**
+     * 
+     * Creates a deep copy of a command node with a different source stack
+     * 
+     * @param sourceNode
+     *            the original command node
+     * @param sourceToResult
+     *            a map storing the original command node as the key and the result command node as the value
+     * @param canUse
+     *            used to check if the player can use the command
+     * @param execute
+     *            the command to execute in place of the old command
+     * @param sourceToResultSuggestion
+     *            a function to convert from the {@link SuggestionProvider} with the original source stack to the {@link SuggestionProvider} with the result source stack
+     * @return the deep copied command node with the new source stack
+     */
+    private static <S, T> CommandNode<T> toResult(CommandNode<S> sourceNode, Map<CommandNode<S>, CommandNode<T>> sourceToResult, S canUse, Command<T> execute,
+            Function<SuggestionProvider<S>, SuggestionProvider<T>> sourceToResultSuggestion)
+    {
+        if (sourceToResult.containsKey(sourceNode))
+            return sourceToResult.get(sourceNode);
+
+        ArgumentBuilder<T, ?> resultBuilder = null;
```

Comment:
The initializer to `null` is unneeded, as variables for reference objects are by default set to `null` and this variable will always have a value (or an exception is thrown).